### PR TITLE
Auth - username/password & token

### PR DESCRIPTION
**Pull request by Jarema**

### CODE REVIEW

File: async-nats/src/lib.rs [hunk 1]
```diff
@@ -980,11 +980,29 @@ pub async fn connect_with_options<A: ToServerAddrs>(
         headers: true,
         no_responders: true,
     };
-    client
-        .sender
-        .send(Command::Connect(connect_info))
-        .await
-        .map_err(|_| io::Error::new(io::ErrorKind::Other, "failed to send connect"))?;
+    match options.auth {
+        Authorization::None => {}
```
Comment (caspervonb) [inline async-nats/src/lib.rs:984]:
```suggestion
```

File: async-nats/src/lib.rs [hunk 2]
```diff
@@ -1397,3 +1415,15 @@ impl<T: ToServerAddrs + ?Sized> ToServerAddrs for &T {
...
+#[derive(Clone)]
+pub(crate) enum Authorization {
+    /// No authentication.
+    None,
+
+    /// Authenticate using a token.
+    Token(String),
+
+    /// Authenticate using a username and password.
+    UsernamePassword(String, String),
```
Comment (caspervonb) [inline async-nats/src/lib.rs:1428]: Awkward name imo

```suggestion
    Credentials(String, String),
```

Or `Basic` (ala http basic authorization)

File: async-nats/src/options.rs [hunk 3]
```diff
@@ -117,6 +119,41 @@ impl ConnectOptions {
...
+    ///
+    /// # Examples
+    /// ```no_run
+    /// # #[tokio::main]
+    /// # async fn main() -> std::io::Result<()> {
+    /// let nc = async_nats::ConnectOptions::with_user_pass("derek".into(), "s3cr3t!".into()).connect("demo.nats.io").await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn with_user_pass(user: String, pass: String) -> Self {
```
Comment (caspervonb) [inline async-nats/src/options.rs:150]:
```suggestion
    pub fn with_user_and_password(user: String, pass: String) -> Self {
```

File: async-nats/src/lib.rs [hunk 4]
```diff
@@ -980,11 +980,29 @@ pub async fn connect_with_options<A: ToServerAddrs>(
...
+        }
+    }
+    connection_handler
+        .connection
+        .write_op(ClientOp::Connect(connect_info))
+        .await?;
+    connection_handler.connection.flush().await?;
+    match connection_handler.connection.read_op().await? {
+        Some(ServerOp::Error(err)) => {
+            return Err(io::Error::new(ErrorKind::InvalidInput, err.to_string()));
```
Comment (caspervonb) [inline async-nats/src/lib.rs:998]: What would the "other" case be here? At this stage there are no subscriptions etc so the only message we'll get are ping/pongs right?

```suggestion
```

**Review by caspervonb — COMMENTED** [review]
One question but overall lgtm and a little bit bike shedding 😁

File: async-nats/src/lib.rs (hunk 4 above)
Comment (Jarema) [reply async-nats/src/lib.rs:998]: yes, but even manually handling `Pong`  only here looks worse than just saying "I don't care, let it be handled by the same logic and flow that handles it in all other cases". IMO this is the safest way, also being the shortest and most consistent.

File: async-nats/src/lib.rs (hunk 2 above)
Comment (Jarema) [reply async-nats/src/lib.rs:1428]: I have to again disagree here, again 😄 

`Credentials` would be easily confused with method that we have to add in the future that implements loading NATS `.creds` files.

**Review by caspervonb — APPROVED** [review]

File: async-nats/src/lib.rs (hunk 2 above)
Comment (caspervonb) [reply async-nats/src/lib.rs:1428]: Option C then 😁

**Pull request merged**